pkg/resources/digitalocean/dns/dns.go

```diff
@@ -0,0 +1,501 @@
+	r.upserts = append(r.upserts, rrset)
+	return r
+}
+
+// Apply adds new records stored in r.additions, updates records stored
+// in r.upserts and deletes records stored in r.removals
+func (r *resourceRecordChangeset) Apply() error {
+	if len(r.additions) > 0 {
+		for _, record := range r.additions {
+			recordCreateRequest := &godo.DomainRecordEditRequest{
+				Name: record.Name(),
+				Data: record.Rrdatas()[0],
+				TTL:  int(record.Ttl()),
+				Type: string(record.Type()),
+			}
+			err := createRecord(r.client, r.zone.Name(), recordCreateRequest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if len(r.removals) > 0 {
+		records, err := getRecords(r.client, r.zone.Name())
+		if err != nil {
+			return err
+		}
+
+		for _, record := range r.removals {
+			var desiredRecord godo.DomainRecord
+			found := false
+			for _, domainRecord := range records {
+				if domainRecord.Name == record.Name() {
+					desiredRecord = domainRecord
+					found = true
+				}
+			}
+			if !found {
+				return errors.New("could not find desired record to remove")
+			}
+
+			err := deleteRecord(r.client, r.zone.Name(), desiredRecord.ID)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	if len(r.upserts) > 0 {
+		records, err := getRecords(r.client, r.zone.Name())
+		if err != nil {
+			return err
+		}
+
+		for _, record := range r.upserts {
+			var desiredRecord godo.DomainRecord
+			found := false
+			for _, domainRecord := range records {
+				if domainRecord.Name == record.Name() {
+					desiredRecord = domainRecord
+					found = true
+				}
+			}
+
+			if !found {
+				return errors.New("could not find desired record to upsert")
+			}
+
+			domainEditRequest := &godo.DomainRecordEditRequest{
+				Name: record.Name(),
+				Data: record.Rrdatas()[0],
+				TTL:  int(record.Ttl()),
+				Type: string(record.Type()),
+			}
+			err := editRecord(r.client, r.zone.Name(), desiredRecord.ID, domainEditRequest)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// IsEmpty returns true if a changeset is empty, false otherwise
+func (r *resourceRecordChangeset) IsEmpty() bool {
+	if len(r.additions) == 0 && len(r.removals) == 0 && len(r.upserts) == 0 {
+		return true
+	}
+
+	return false
+}
+
+// ResourceRecordSet returns the associated resourceRecordSets of a changset
+func (r *resourceRecordChangeset) ResourceRecordSets() dnsprovider.ResourceRecordSets {
+	return r.rrsets
+}
+
+// listDomains returns a list of godo.Domain
+func listDomains(c *godo.Client) ([]godo.Domain, error) {
+	// TODO (andrewsykim): pagination in ListOptions
+	domains, resp, err := c.Domains.List(context.TODO(), &godo.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != 200 {
+		respData, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("error reading response body: %v", err)
+		}
+
+		return nil, fmt.Errorf("received non 200 status code: %d from api: %v",
+			resp.StatusCode, string(respData))
+	}
```

Comment:
I think you may have mistaken the check above for `if resp.StatusCode == 200` but it's actually `if resp.StatusCode != 200`. So we only do err handling when the status code is not 200, otherwise we just return whatever the API returned